pkg/render/egressgateway/egressgateway.go

```diff
@@ -0,0 +1,438 @@
+
+package egressgateway
+
+import (
+	"fmt"
+	"strings"
+
+	ocsv1 "github.com/openshift/api/security/v1"
+	operatorv1 "github.com/tigera/operator/api/v1"
+	"github.com/tigera/operator/pkg/components"
+	"github.com/tigera/operator/pkg/ptr"
+	"github.com/tigera/operator/pkg/render"
+	rmeta "github.com/tigera/operator/pkg/render/common/meta"
+	"github.com/tigera/operator/pkg/render/common/podsecuritypolicy"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	policyv1beta1 "k8s.io/api/policy/v1beta1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	egwPortName               = "health"
+	DefaultEGWVxlanPort int   = 4790
+	DefaultEGWVxlanVNI  int   = 4097
+	DefaultHealthPort   int32 = 8080
+)
+
+func EgressGateway(
+	config *Config,
+) render.Component {
+	return &component{
+		config: config,
+	}
+}
+
+type component struct {
+	config *Config
+}
+
+// Config contains all the config information EgressGateway needs to render component.
+type Config struct {
+	PullSecrets  []*corev1.Secret
+	Installation *operatorv1.InstallationSpec
+	OSType       rmeta.OSType
+	EgressGW     *operatorv1.EgressGateway
+
+	egwImage          string
+	EgressGWVxlanVNI  int
+	EgressGWVxlanPort int
+
+	Openshift bool
+	// Whether or not the cluster supports pod security policies.
+	UsePSP bool
+}
+
+func (c *component) ResolveImages(is *operatorv1.ImageSet) error {
+	reg := c.config.Installation.Registry
+	path := c.config.Installation.ImagePath
+	prefix := c.config.Installation.ImagePrefix
+
+	if c.config.OSType != c.SupportedOSType() {
+		return fmt.Errorf("Egress Gateway is supported only on %s", c.SupportedOSType())
+	}
+
+	var err error
+	c.config.egwImage, err = components.GetReference(components.ComponentEgressGateway, reg, path, prefix, is)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (c *component) SupportedOSType() rmeta.OSType {
+	return rmeta.OSTypeLinux
+}
+
+func (c *component) Objects() ([]client.Object, []client.Object) {
+	objects := []client.Object{}
+	objects = append(objects, c.egwServiceAccount())
+	objects = append(objects, c.egwDeployment())
+	if c.config.UsePSP {
+		objects = append(objects, c.egwPodSecurityPolicy())
+		objects = append(objects, c.egwClusterRole())
+		objects = append(objects, c.egwClusterRoleBinding())
+	} else if c.config.Openshift {
+		objects = append(objects, c.egwSecurityContextConstraints())
+	}
+	return objects, nil
+}
+
+func (c *component) Ready() bool {
+	return true
+}
+
+func (c *component) egwDeployment() *appsv1.Deployment {
+	d := appsv1.Deployment{
+		TypeMeta: metav1.TypeMeta{Kind: "Deployment", APIVersion: "apps/v1"},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      c.config.EgressGW.Name,
+			Namespace: c.config.EgressGW.Namespace,
+			Labels:    c.config.EgressGW.Spec.Template.Metadata.Labels,
```

Comment:
Labels are handled by the override mechanism as well, so shouldn't be done here IIUC.

Reply:
I think you missed this comment - Labels and other metadata are handled here: https://github.com/tigera/operator/blob/ddbaa22cd1b351564e952b851bc1b6b5d0aeae0d/pkg/render/common/components/components.go#L41-L50

Is there a reason that isn't sufficient for EGW?

Reply:
For now, ```GetMetadata``` returns nil for EGW. To be fixed soon.